File: 01-Programming-Languages/C++/Algorithms/Sorting/merge_sort.cpp
/**
 * Problem: Merge Sort
 * Given an array of integers, sort the array in ascending order using the merge sort algorithm.
 * Merge sort is a divide-and-conquer algorithm that divides the array into two halves,
 * recursively sorts them, and then merges the two sorted halves.
 *
 * Approach:
 * - Divide the array into two halves
 * - Recursively sort each half
 * - Merge the two sorted halves into a single sorted array
 * - The merge process compares elements from each half and combines them in sorted order
 *
 * Time Complexity: O(n log n) - divides array log n times, each level does O(n) work
 * Space Complexity: O(n) - temporary array needed for merging
 *
 * Example:
 * Input: [38, 27, 43, 3, 9, 82, 10]
 * Output: [3, 9, 10, 27, 38, 43, 82]
 */

#include <iostream>
#include <vector>

/**
 * Merges two sorted subarrays into a single sorted array
 *
 * @param arr Reference to the vector containing the subarrays to merge
 * @param left Starting index of the left subarray
 * @param mid Ending index of the left subarray (mid+1 is start of right subarray)
 * @param right Ending index of the right subarray
 */
void merge(std::vector<int>& arr, int left, int mid, int right) {
    // Calculate sizes of the two subarrays
    int n1 = mid - left + 1;
    int n2 = right - mid;

    // Create temporary vectors
    std::vector<int> leftArr(n1);
    std::vector<int> rightArr(n2);

    // Copy data to temporary vectors
    for (int i = 0; i < n1; i++) {
        leftArr[i] = arr[left + i];
    }
    for (int j = 0; j < n2; j++) {
        rightArr[j] = arr[mid + 1 + j];
    }

    // Merge the temporary vectors back into arr[left..right]
    int i = 0;   // Index for leftArr
    int j = 0;   // Index for rightArr
    int k = left; // Index for arr

    while (i < n1 && j < n2) {
        if (leftArr[i] <= rightArr[j]) {
            arr[k] = leftArr[i];
            i++;
        } else {
            arr[k] = rightArr[j];
            j++;
        }
        k++;
    }

    // Copy remaining elements of leftArr, if any
    while (i < n1) {
        arr[k] = leftArr[i];
        i++;
        k++;
    }

    // Copy remaining elements of rightArr, if any
    while (j < n2) {
        arr[k] = rightArr[j];
        j++;
        k++;
    }
}

/**
 * Main merge sort function that recursively sorts the array
 *
 * @param arr Reference to the vector to be sorted
 * @param left Starting index
 * @param right Ending index
 */
void mergeSort(std::vector<int>& arr, int left, int right) {
    if (left < right) {
        // Find the middle point
        int mid = left + (right - left) / 2;

        // Sort first and second halves
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);

        // Merge the sorted halves
        merge(arr, left, mid, right);
    }
}

// Helper function to print a vector
void printVector(const std::vector<int>& vec) {
    std::cout << "[";
    for (size_t i = 0; i < vec.size(); i++) {
        std::cout << vec[i];
        if (i < vec.size() - 1) {
            std::cout << ", ";
        }
    }
    std::cout << "]" << std::endl;
}

// Example usage
int main() {
    // Example input
    std::vector<int> arr = {38, 27, 43, 3, 9, 82, 10};

    std::cout << "Original array: ";
    printVector(arr);

    // Sort the array using merge sort
    mergeSort(arr, 0, arr.size() - 1);

    std::cout << "Sorted array:   ";
    printVector(arr);

    // Additional test cases
    std::vector<int> test1 = {5, 2, 4, 6, 1, 3};
    std::cout << "\nTest 1:\n";
    std::cout << "Original: ";
    printVector(test1);
    mergeSort(test1, 0, test1.size() - 1);
    std::cout << "Sorted:   ";
    printVector(test1);

    std::vector<int> test2 = {1};
    std::cout << "\nTest 2 (single element):\n";
    std::cout << "Original: ";
    printVector(test2);
    mergeSort(test2, 0, test2.size() - 1);
    std::cout << "Sorted:   ";
    printVector(test2);

    std::vector<int> test3 = {};
    std::cout << "\nTest 3 (empty array):\n";
    std::cout << "Original: ";
    printVector(test3);
    mergeSort(test3, 0, test3.size() - 1); // Safe to call with empty vector
    std::cout << "Sorted:   ";
    printVector(test3);

    return 0;
}

File: 01-Programming-Languages/Python/Algorithms/Sorting/merge_sort.py
"""
Problem: Merge Sort
Given an array of integers, sort the array in ascending order using the merge sort algorithm.
Merge sort is a divide-and-conquer algorithm that divides the array into two halves,
recursively sorts them, and then merges the two sorted halves.

Approach:
- Divide the array into two halves
- Recursively sort each half
- Merge the two sorted halves into a single sorted array
- The merge process compares elements from each half and combines them in sorted order

Time Complexity: O(n log n) - divides array log n times, each level does O(n) work
Space Complexity: O(n) - temporary array needed for merging

Example:
Input: [38, 27, 43, 3, 9, 82, 10]
Output: [3, 9, 10, 27, 38, 43, 82]
"""

from typing import List

def merge(arr: List[int], left: int, mid: int, right: int) -> None:
    """
    Merges two sorted subarrays into a single sorted array

    Args:
        arr: Array containing the subarrays to merge
        left: Starting index of the left subarray
        mid: Ending index of the left subarray (mid+1 is start of right subarray)
        right: Ending index of the right subarray
    """
    # Calculate sizes of the two subarrays
    n1 = mid - left + 1
    n2 = right - mid

    # Create temporary arrays
    left_arr = [0] * n1
    right_arr = [0] * n2

    # Copy data to temporary arrays
    for i in range(n1):
        left_arr[i] = arr[left + i]
    for j in range(n2):
        right_arr[j] = arr[mid + 1 + j]

    # Merge the temporary arrays back into arr[left..right]
    i = 0   # Index for left_arr
    j = 0   # Index for right_arr
    k = left  # Index for arr

    while i < n1 and j < n2:
        if left_arr[i] <= right_arr[j]:
            arr[k] = left_arr[i]
            i += 1
        else:
            arr[k] = right_arr[j]
            j += 1
        k += 1

    # Copy remaining elements of left_arr, if any
    while i < n1:
        arr[k] = left_arr[i]
        i += 1
        k += 1

    # Copy remaining elements of right_arr, if any
    while j < n2:
        arr[k] = right_arr[j]
        j += 1
        k += 1

def merge_sort(arr: List[int], left: int, right: int) -> None:
    """
    Main merge sort function that recursively sorts the array

    Args:
        arr: Array to be sorted
        left: Starting index
        right: Ending index
    """
    if left < right:
        # Find the middle point
        mid = left + (right - left) // 2

        # Sort first and second halves
        merge_sort(arr, left, mid)
        merge_sort(arr, mid + 1, right)

        # Merge the sorted halves
        merge(arr, left, mid, right)

# Helper function to print an array
def print_array(arr: List[int]) -> None:
    """Prints the array in a readable format"""
    print(f"[{', '.join(map(str, arr))}]")

# Example usage
if __name__ == "__main__":
    # Example input
    arr = [38, 27, 43, 3, 9, 82, 10]

    print("Original array:")
    print_array(arr)

    # Sort the array using merge sort
    merge_sort(arr, 0, len(arr) - 1)

    print("Sorted array:")
    print_array(arr)

    # Additional test cases
    print("\nTest 1:")
    test1 = [5, 2, 4, 6, 1, 3]
    print("Original:")
    print_array(test1)
    merge_sort(test1, 0, len(test1) - 1)
    print("Sorted:")
    print_array(test1)

    print("\nTest 2 (single element):")
    test2 = [1]
    print("Original:")
    print_array(test2)
    merge_sort(test2, 0, len(test2) - 1)
    print("Sorted:")
    print_array(test2)

    print("\nTest 3 (empty array):")
    test3 = []
    print("Original:")
    print_array(test3)
    merge_sort(test3, 0, len(test3) - 1)  # Safe to call with empty array
    print("Sorted:")
    print_array(test3)